Documentation/Contributing.md

```diff
@@ -0,0 +1,350 @@
+repositories
+x86_64-apple-macosx
+debug.yaml
+manifest.db
+workspace-state.json 
+```
+Binaries (in the example above) are in `x86_64-apple-macosx/`.  
+If you need to build the generated binaries, run `swift-build` in inside `.build/`:
+```bash
+./.build/x86_64-apple-macosx/debug/swift-build
+```
+
+## Running Tests
+### Run all tests.
+```bash
+$> swift test
+```
+### Run a single test.
+```bash
+$> swift test --filter PackageGraphTests.DependencyResolverTests/testBasics
+```
+Or another example, to run tests for the test targets BuildTests and WorkspaceTests, but skip some test cases.
+```bash
+$> swift test --filter BuildTests --skip BuildPlanTests --filter WorkspaceTests --skip InitTests
+```
+Note: PackageDescription v4 is not available when developing using this method.
+
+This method can also used be used for performance testing. Use the following command run SwiftPM's performance tests:
+
+```bash
+$> export TSC_ENABLE_PERF_TESTS=1
+$> swift test -c release --filter PerformanceTests
+```
+
+To test SwiftPM functionality using the newly created binaries, you can invoke them from the `.build` directory. For example:
+
+```bash
+$> cd /tmp && mkdir hello && cd hello
+$> /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-package init
+$> /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-build
+```
+
+## Testing on Linux with Docker
+For contributors on macOS who need to test on Linux, install Docker and use the
+following commands:
+
+```bash
+$> Utilities/Docker/docker-utils build # will build an image with the latest Swift snapshot
+$> Utilities/Docker/docker-utils bootstrap # will bootstrap SwiftPM on the Linux container
+$> Utilities/Docker/docker-utils run bash # to run an interactive Bash shell in the container
+$> Utilities/Docker/docker-utils swift-build # to run swift-build in the container
+$> Utilities/Docker/docker-utils swift-test # to run swift-test in the container
+$> Utilities/Docker/docker-utils swift-run # to run swift-run in the container
+```
+
+## Creating Pull Requests
+1. Fork: https://github.com/apple/swift-package-manager
+2. Clone a working copy of your fork
+3. Create a new branch
+4. Make your code changes
+5. Try to keep your changes (when possible) below 200 lines of code.
+6. We use [SwiftFormat](https://www.github.com/nicklockwood/SwiftFormat) to enforce code style. Please install and run SwiftFormat before submitting your PR.
+7. Commit (include the Radar link or JIRA issue id in the commit message if possible and a description your changes). Try to have only 1 commit in your PR (but, of course, if you add changes that can be helpful to be kept aside from the previous commit, make a new commit for them).
+8. Push the commit / branch to your fork
+9. Make a PR from your fork / branch to `apple: main`
+10. While creating your PR, make sure to add one of the [PR Templates](https://github.com/apple/swift-package-manager/tree/main/.github) in your PR message depending on the reasoning of your PR.
```

Comment:
While creating your PR, make sure to follow the PR Template providing information about the motivation and highlighting the changes.